### PR TITLE
Cluster - Add app-version to the Member information

### PR DESCRIPTION
**Pull request by zbynek001**
Add app-version to the Member information, (Migrated from https://github.com/akka/akka/pull/29546)
will be used in rolling update features
configured with akka.cluster.app-version

### CODE REVIEW

File: build.fsx [hunk 1]
```diff
@@ -545,6 +545,7 @@ Target "Protobuf" <| fun _ ->
                 |> append (sprintf "-I=%s" (__SOURCE_DIRECTORY__ @@ "/src/protobuf/common") )
                 |> append (sprintf "--csharp_out=internal_access:%s" (__SOURCE_DIRECTORY__ @@ destinationPath))
                 |> append "--csharp_opt=file_extension=.g.cs"
+                |> append "--experimental_allow_proto3_optional"
```
Comment (zbynek001) [inline build.fsx:548]: needed for `optional` proto members

File: build.ps1 [hunk 2]
```diff
@@ -35,7 +35,7 @@ $DotNetVersion = "3.1.105";
 $DotNetInstallerUri = "https://dot.net/v1/dotnet-install.ps1";
 $NugetVersion = "4.3.0";
 $NugetUrl = "https://dist.nuget.org/win-x86-commandline/v$NugetVersion/nuget.exe"
-$ProtobufVersion = "3.4.0"
+$ProtobufVersion = "3.13.0"
```
Comment (zbynek001) [inline build.ps1:38]: needed for `optional` proto members

File: src/core/Akka/Util/AkkaVersion.cs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+    ///  plain alphanumeric if it couldn't be parsed as a number.
+    ///
+    ///  It may also have a qualifier at the end for 2 or 3 digit version numbers such as "1.2-RC1".
+    ///  For 1 digit with qualifier, 1-RC1, it is interpreted as plain alphanumeric.
+    ///
+    ///  It has support for https://github.com/dwijnand/sbt-dynver format with `+` or
+    ///  `-` separator. The number of commits from the tag is handled as a numeric part.
+    ///  For example `1.0.0+3-73475dce26` is less than `1.0.10+10-ed316bd024` (3 < 10).
+    /// </summary>
+    public class AkkaVersion : IComparable<AkkaVersion>, IEquatable<AkkaVersion>
```
Comment (zbynek001) [inline src/core/Akka/Util/AkkaVersion.cs:32]: Not sure about the name. Should be called just `Version`, but that would conflict everywhere with `System.Version`

File: build.ps1 (hunk 2 above)
Comment (Aaronontheweb) [reply build.ps1:38]: Got it

File: build.fsx (hunk 1 above)
Comment (Aaronontheweb) [reply build.fsx:548]: LGTM

File: src/core/Akka.API.Tests/CoreAPISpec.ApproveCluster.approved.txt [hunk 4]
```diff
@@ -73,6 +73,7 @@ namespace Akka.Cluster
             public CurrentClusterState() { }
             public CurrentClusterState(System.Collections.Immutable.ImmutableSortedSet<Akka.Cluster.Member> members, System.Collections.Immutable.ImmutableHashSet<Akka.Cluster.Member> unreachable, System.Collections.Immutable.ImmutableHashSet<Akka.Actor.Address> seenBy, Akka.Actor.Address leader, System.Collections.Immutable.ImmutableDictionary<string, Akka.Actor.Address> roleLeaderMap) { }
             public System.Collections.Immutable.ImmutableHashSet<string> AllRoles { get; }
+            public bool HasMoreThanOneAppVersion { get; }
```
Comment (Aaronontheweb) [inline src/core/Akka.API.Tests/CoreAPISpec.ApproveCluster.approved.txt:76]: Ran into this issue the other day with a large user cluster - looks like a good addition

File: src/core/Akka.Cluster.Tests/ClusterConfigSpec.cs [hunk 5]
```diff
@@ -45,6 +46,7 @@ public void Clustering_must_be_able_to_parse_generic_cluster_config_elements()
             settings.MinNrOfMembers.Should().Be(1);
             settings.MinNrOfMembersOfRole.Should().Equal(ImmutableDictionary<string, int>.Empty);
             settings.Roles.Should().BeEquivalentTo(ImmutableHashSet<string>.Empty);
+            settings.AppVersion.Should().Be(AkkaVersion.Zero);
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster.Tests/ClusterConfigSpec.cs:49]: LGTM

File: src/core/Akka.Cluster/Configuration/Cluster.conf [hunk 6]
```diff
@@ -77,6 +77,23 @@ akka {
...
+    # plain alphanumeric if it couldn't be parsed as a number.
+    #
+    # It may also have a qualifier at the end for 2 or 3 digit version numbers such
+    # as "1.2-RC1".
+    # For 1 digit with qualifier, 1-RC1, it is interpreted as plain alphanumeric.
+    #
+    # It has support for https://github.com/dwijnand/sbt-dynver format with `+` or
+    # `-` separator. The number of commits from the tag is handled as a numeric part.
+    # For example `1.0.0+3-73475dce26` is less than `1.0.10+10-ed316bd024` (3 < 10).
+    app-version = "0.0.0"
```
Comment (Aaronontheweb) [inline src/core/Akka.Cluster/Configuration/Cluster.conf:95]: So the App Version is used to specify the version of the user's software, not Akka.NET... so to take advantage of this, users will need to populate it inside their HOCONs. We'll need to add this to our cluster documentation.

File: src/core/Akka/Util/AkkaVersion.cs (hunk 3 above)
Comment (Aaronontheweb) [reply src/core/Akka/Util/AkkaVersion.cs:32]: I can rename it `AppVersion` after merge - that way it wouldn't be confused with the Akka.NET version, which is what I initially thought when I saw the name.

**Review by Aaronontheweb — APPROVED** [review]
LGTM

**Pull request merged**